# Bitwise **NOT** Operator: `~x`
---

- `~x` = 1's complement of ***x***


1. Find Complement of ***x***: 
    - flip all bits (includin sign)
    - get intermediate result
2. Flip inetermediate result and add/subtract 1 (depending on sign)
3. Result = Complement 


- Python Does Not Use A *Sign Bit*
    - `-13 = -0b1101`
    - stored externally 

### Positive Start 
$\begin{array}{c|c|llll|c}
{step}&{sign}&8&4&3&1&decimal\\
\hline
\hline
{original}&+&1&1&0&1&+13\\
{flip}\\
\hline
{comp1}&-&0&0&1&0&-2\\
{flip}&+&1&1&0&1&+13\\
{add 1}&&&&+&1\\
\hline
{comp2}&-&1&1&1&0&-14\\
\end{array}$

### Negative Start 
$\begin{array}{c|c|llll|c}
{step}&{sign}&8&4&3&1&decimal\\
\hline
\hline
{original}&-&1&1&1&0&-14\\
{flip}\\
\hline
{comp1}&+&0&0&0&1&+1\\
{flip}&-&1&1&1&0&-14\\
{minus 1}&&&&-&1\\
\hline
{comp2}&+&1&1&0&1&+13\\
\end{array}$

In [1]:
def complement_1(decimal):
    if decimal >= 0: 
        s = '0'
        b = bin(decimal)[2:]
    else: 
        s = '1' 
        b = bin(decimal)[3:]
    ones,twos = "",""
    print(f"Original Binary: {b} \n Sign: {s}")

    def flip(c): 
        if c == '1': return '0'
        elif c == '0': return '1'
        else: return "error"

    # 1's Complement 
    for i in range(len(b)): 
        ones += flip(b[i])
    s = flip(s)
    print(f"Complement 1: {ones} \n Sign: {s}")

    # 2's Complement
    twos = list(ones)
    for i in range(len(twos)):
        twos[i] = flip(twos[i])
    s = flip(s)
    print(f"Complement 2 FLIP PREP: {twos} \n Sign: {s}")

    # Add 1
    addy = '1'.zfill(len(b))
    result = ''
    carry = 0 
    if s == '0':
        for i in range(len(twos) - 1, -1,-1): 
            temp = carry 
            temp += 1 if twos[i] == '1' else 0 
            temp += 1 if addy[i] == '1' else 0 
            result = ('1' if temp % 2 == 1 else '0') + result 

            carry = 0 if temp < 2 else 1 
        if carry != 0: 
            result = '1' + result 
        print(f"POSITIVE! -> ~{decimal} = {result}")
    
    #PICKUP HERE 
    else: 
        for i in range(len(twos)-1, -1,-1): 
            temp = carry 
            if twos[i] == '1':
                twos[i] == 0
                break
            else: 
                if twos[i-1] == '1':
                    twos[i] == '1'
                    twos[i-1] == '0'
                    break
                else: 
                    i -= 1
        print(f"NEGATIVE! ~{decimal} = {''.join(twos)}")
    return twos


pos1 = complement_1(13)
neg1 = complement_1(-14)
print(pos1, neg1)

Original Binary: 1101 
 Sign: 0
Complement 1: 0010 
 Sign: 1
Complement 2 FLIP PREP: ['1', '1', '0', '1'] 
 Sign: 0
POSITIVE! -> ~13 = 1110
Original Binary: 1110 
 Sign: 1
Complement 1: 0001 
 Sign: 0
Complement 2 FLIP PREP: ['1', '1', '1', '0'] 
 Sign: 1
NEGATIVE! ~-14 = 1110
['1', '1', '0', '1'] ['1', '1', '1', '0']


In [2]:
def flip(c): 
    return '1' if (c == '0') else '0'

def complements(x): 
    b = bin(x)[2:]
    b = b.zfill(8)
    n = len(b)
    print(f"Original Binary: {b}")
    ones,twos = "",""

    # 1's Complement: Flip Every Bit 
    for i in range(n): 
        ones += flip(b[i])
    print(f"1's Complement:  {ones}\n")

    # 2's Complement: Find First 1 -> Then Flip 
    ones = list(ones.strip("")) 
    twos = list(ones)

    for i in range(n-1, -1, -1): #index and reverse 
        if ones[i] == '1': 
            twos[i] = '0'
        else: 
            twos[i] = '1'
            break
    i -= 1
    if i == -1: 
        twos.insert(0,'1')


    return ones, twos

num = 12
complements(num)

Original Binary: 00001100
1's Complement:  11110011



(['1', '1', '1', '1', '0', '0', '1', '1'],
 ['1', '1', '1', '1', '0', '1', '0', '0'])

In [3]:
a,b = 5,21
na, nb = bin(a)[2:], bin(b)[2:]

print("binary a: {}, {}\n OG:  {} \nNOT: {} \nSUM: {} \n".format(bin(a),na,a,(~a), a+~a))
print("binary b: {}, {}\n OG:  {} \nNOT: {} \nSUM: {}".format(bin(b),nb,b,(~b), b+~b))


binary a: 0b101, 101
 OG:  5 
NOT: -6 
SUM: -1 

binary b: 0b10101, 10101
 OG:  21 
NOT: -22 
SUM: -1


In [4]:
def check_bitwise_not(list): 
    for i in range(len(list)):
        print("binary {}: {}".format(list[i],bin(list[i])))
        print(" NOT: {}".format(~list[i]))

not_list = [15,-15]
check_bitwise_not(not_list)

binary 15: 0b1111
 NOT: -16
binary -15: -0b1111
 NOT: 14


### Numbers Stored in Binary Format 
 - __ ________ Leftmost Bit Determines Sign
    - 0 -> Positive
    - 1 -> Negative
- x = 13 -> binary_x = **0** 1101
    - leftmost reserved for sign, `0 = positive`
> #### Original: `~13` or  `0 | 1101`
> #### 1's Complement: `1 | 0010` flip everything including sign 
> #### 2's Complement: `1 | 1101` flip 1's complement - except sign 
>> #### ADD ONE to 2's Complement: ($1+1=10$)
>>>
>>> $1 | 1110= \begin{vmatrix} -&0&0&1&0 \cr +&1&1&0&1 \cr &&&+&1 \end{vmatrix}$

### Positive Start 
$\begin{array}{c|c|llll|c}
{step}&{sign}&8&4&3&1&decimal\\
\hline
\hline
{original}&+&1&1&0&1&+13\\
{flip}\\
\hline
{comp1}&-&0&0&1&0&-2\\
{flip}&+&1&1&0&1&+13\\
{add 1}&&&&+&1\\
\hline
{comp2}&-&1&1&1&0&-14\\
\end{array}$

### Negative Start 
$\begin{array}{c|c|llll|c}
{step}&{sign}&8&4&3&1&decimal\\
\hline
\hline
{original}&-&1&1&1&0&-14\\
{flip}\\
\hline
{comp1}&+&0&0&0&1&+1\\
{flip}&-&1&1&1&0&-14\\
{minus 1}&&&&-&1\\
\hline
{comp2}&+&1&1&0&1&+13\\
\end{array}$


### 1's Complement: Flipping Bits 
### 2's Complement: Add 1 to 1's Complement 
- negative number stored as two's complement 
- ADD 1: `1 + 1 = 10` -> `1101 + 1 = 1110`

In [5]:
not_zero, not_one = ~0, ~1
print(f"NOT Zero: ~0 = {not_zero} \nNOT One: ~1 = {not_one}")

NOT Zero: ~0 = -1 
NOT One: ~1 = -2


In [6]:
pos = 0b1101
neg = -0b1101
print(f"Positive: {pos} \nNegative: {neg}")

Positive: 13 
Negative: -13
